vivid: replace theme content from yaml with json

arunoruto opened the pull request:
### Description

As per the discussion in #7994 and in https://github.com/nix-community/stylix/pull/1938, this PR implements a `home.sessionVariables.LS_COLORS` and assigns it a value at evaluation by running vivid with `pkgs.runCommand`.

Since `home.sessionVariables.LS_COLORS` supersedes the shell integrations (https://github.com/nix-community/home-manager/issues/7994#issuecomment-3403827062), they have been removed.

Partially addresses the issue #7994. 
Values like `1e2030` are treated as strings in Nix, but since they have a single non-digit character, don't get enclosed in single quotes. This leads to the problem that such hex codes are treated as numbers in scientific notation.
Since json is just a yaml subset (in simple terms, see [hackers news](https://news.ycombinator.com/item?id=31406473)), we can use `builtins.toJSON`. JSON has much better support in Nix, making it more robust for such a usecase!


### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://nix-community.github.io/home-manager/#sec-guidelines

-->

- [x] Change is backwards compatible.

- [x] Code formatted with `nix fmt` or
      `nix-shell -p treefmt nixfmt deadnix keep-sorted --run treefmt`.

- [ ] Code tested through `nix run .#tests -- test-all` or
      `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [ ] Commit messages are formatted like

  ```
  {component}: {description}

  {long description}
  ```

  See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/a51598236f23c89e59ee77eb8e0614358b0e896c/modules/programs/lesspipe.nix#L11).
  - [ ] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)
  - [ ] Basic tests added. See [Tests](https://nix-community.github.io/home-manager/index.xhtml#sec-tests)

- If this PR adds an exciting new feature or contains a breaking change.
  - [ ] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)

aguirre-matteo commented on modules/programs/vivid.nix:
```suggestion
      type = with types; nullOr (enum [
        "8-bit"
        "24-bit"
      ]);
      default = null;
      example = "8-bit";
```

aguirre-matteo replied:
Just to make the option a bit more secure to use. I defaulted the option to `null` instead of `24-bit` so by default the command would not set the `-m` flag, so in the future if the default value changes we'll still be using `-m 24-bit`, and that could lead to unexpected behavior.

aguirre-matteo commented on modules/programs/vivid.nix:
```diff
@@ -116,16 +110,27 @@ in
+        })";
+    })
+    (mkIf cfg.enable {
+      home.sessionVariables = mkIf (cfg.activeTheme != "") { VIVID_THEME = cfg.activeTheme; };
```
Good overall, but I'm pretty sure we'll face some error during the tests check. You should also add assertions that ensure `cfg.themes` contains `cfg.activeTheme` since otherwise the user will face an ugly error message at evaluation time. You should also set an assertion ensuring the user sets the `cfg.activeTheme` option, since otherwise if we give `vivid` no arguments it will throw and error and the derivation will fail.

trueNAHO replied:
> Good overall, but I'm pretty sure we'll face some error during the tests check. You should also add assertions that ensure `cfg.themes` contains `cfg.activeTheme` since otherwise the user will face an ugly error message at evaluation time. You should also set an assertion ensuring the user sets the `cfg.activeTheme` option, since otherwise if we give `vivid` no arguments it will throw an error and the derivation will fail.

The problem is that we are currently always treating `cfg.activeTheme` to be defined as a `cfg.themes` YAML, which is too restrictive. My previous comment addresses this.

> I'm a bit busy right now, so I can't review this part by part, but I'll share the code I was working on yesterday so you can adapt it.

Your proposal makes major changes to the guarding and makes default values more restrictive. Without separate discussions this is too much at once to discuss.

> ```nix
>     let
>       vividThemes = lib.mapAttrs' (
>         name: value:
>         lib.nameValuePair name {
>           source = if lib.isAttrs value then jsonFormat.generate "${name}.yml" value else value;
>         }
>       ) cfg.themes;
>     in
> ```

Centralizing the `jsonFormat.generate` calls seems like a good addition.

trueNAHO commented on modules/programs/vivid.nix:
```diff
@@ -116,16 +110,27 @@ in
+        "$(cat ${
+          pkgs.runCommand "ls-colors" {
+            nativeBuildInputs = [ cfg.package ];
+          } "vivid ${colorMode} generate ${themePath} > $out"
```
Nitpick:

```suggestion
          } "vivid ${colorMode} generate ${themePath} >$out"
```

arunoruto commented on modules/programs/vivid.nix:
```diff
@@ -134,21 +139,13 @@ in
+            { source = pkgs.writeText "${name}.json" (builtins.toJSON value); }
+          else
+            { source = value; }
+        )
```
Hope the small comment is enough

arunoruto commented on modules/programs/vivid.nix:
```diff
@@ -7,32 +7,26 @@
+  meta.maintainers = [
+    lib.hm.maintainers.aguirre-matteo
+    lib.maintainers.arunoruto
+  ];
```
Seems like they do not merge it, at least not in my config:
```nix
nix-repl> hm.config.lib.maintainers.arunoruto
error: attribute 'arunoruto' missing
       at «string»:1:1:
            1| hm.config.lib.maintainers.arunoruto
```

But aguirre-matteo is:
```nix
nix-repl> hm.config.lib.maintainers.aguirre-matteo
{
  github = "aguirre-matteo";
  ...
}
```

aguirre-matteo replied:
@arunoruto Add yourself to `modules/lib/maintainers.nix`.

arunoruto replied:
I am a bit confused, since the [docs](https://nix-community.github.io/home-manager/index.xhtml#_add_yourself_as_a_module_maintainer) state it differently:

## Add yourself as a module maintainer

Every new module must include a named maintainer using the meta.maintainers attribute. If you are a user of a module that currently lacks a maintainer then please consider adopting it.

**If you are present in the nixpkgs maintainer list then you can use that entry. If you are not then you can add yourself to modules/lib/maintainers.nix in the Home Manager project.**

As a maintainer you are expected to respond to issues and pull-requests associated with your module.

Maintainers are encouraged to join the IRC or Matrix channel and participate when they have opportunity.

arunoruto commented on modules/programs/vivid.nix:
```diff
@@ -116,16 +110,27 @@ in
-
-      home.sessionVariables = mkIf (cfg.activeTheme != null) { VIVID_THEME = cfg.activeTheme; };
+  config = mkMerge [
+    (mkIf (cfg.enable || cfg.themes != { }) {
```
I removed it due to the comment of @aguirre-matteo: https://github.com/nix-community/home-manager/issues/7994#issuecomment-3403499199

So we would have one option, which would set the session variable `LS_COLORS` without installing the package and setting the files in the `$XDG_CONFIG_HOME/vivid` folder.

arunoruto commented on modules/programs/vivid.nix:
```diff
@@ -116,16 +110,27 @@ in
-
-      home.sessionVariables = mkIf (cfg.activeTheme != null) { VIVID_THEME = cfg.activeTheme; };
+  config = mkMerge [
+    (mkIf (cfg.enable || cfg.themes != { }) {
```
> Sorry but that makes no sense to me. We shouldn't set the `LS_COLORS` env variable unless users explicitly set they want vivid to do so.

His idea was also to include a `enableSessionVariables` option, so `didWeDeclareTheYamlAttributeSet == cfg.enableSessionVariables && cfg.themes != { }`. This wouldn't mess with the users environment. I guess this is my fault, since I excluded it due to your comment about not needing additional options.

aguirre-matteo replied:
> I guess this is my fault, since I excluded it due to your comment about not needing additional options.

The point I was trying to make is that the main purpose of the module is to set the `LS_COLORS` variable through vivid, and I barely can think about a use case where users configure `programs.vivid` but don't want to use it in their shells. Yet, if you think such option adds some value feel free to implement it, but I don't see it as necessary.

arunoruto replied:
> The point I was trying to make is that the main purpose of the module is to set the `LS_COLORS` variable through vivid, and I barely can think about a use case where users configure `programs.vivid` but don't want to use it in their shells.

I think I get your point and the additional option would make more sense, if we still had shell integrations, so we can let the user choose if they want to enable theming via session variables or shell integrations.

What @trueNAHO is proposing, is to still have an additional option of just enabling that one single theme and that's it. Not additional saving of themes in `~/.config/vivid/themes` folder or setting the `VIVID_THEME` variable. None of this is technically needed to just apply the `activeTheme`. This would indeed reduce the amount of "bloat" when using Stylix, since we just care for one theme, namely the generated stylix-theme.

Maybe we're trying to over-engineer something that should be kept simple; the end user won't notice it much either way. I guess keeping it simple would be the way to go, i.e., having one `enable` option, which would make the `mkMerge` redundant.

And please don't get me wrong, I like the approach of fine-tuning the configuration, allowing the user to keep only the essential parts. But the question is, is it worth the additional "complexity" in the end?

arunoruto commented:
> I think we are discussing too much things at the same time in this PR, so I propose we split it in a series of PRs to clean the commit history and/or better discuss them, and since they are unrelated to each other. We could split them in:
> 
> 1- Replacing `yamlFormat` to `jsonFormat` to address the scientific notation issue.
> 
> 2- Introducing `programs.vivid.theme`, deprecating the `themes` and `activeTheme` options.
> 
> 3- Change the type of `colorMode` to and enum.
> 
> 4- Generating `LS_COLORS` at evaluation time. We could keep this in this same PR or start a new one of you want to make it more clean. I think it would be wise to have more opinions here about the use of the OR operator and the other details.

Agreed, I changed the contents to just reflect point 1. I guess we can do 4 in a separate PR to get more feedback.

Btw, I am not sure why the tests are failing... The content is the same, but it says they are different:
```diff
--- actual
+++ expected
@@ -1 +1 @@
-{"colors":{"primary":"00aaff","secondary":"ff00aa"},"core":{"directory":{"foreground":"primary"},"executable-file":{"font-style":"bold","foreground":"secondary"}}}
\ No newline at end of file
+{"colors":{"primary":"00aaff","secondary":"ff00aa"},"core":{"directory":{"foreground":"primary"},"executable-file":{"font-style":"bold","foreground":"secondary"}}}
```

I will rewrite the PR title and contents.

arunoruto commented:
> Do `cp /nix/store/fnzcfj3vxnndbdnl15mcb117ilzyms2i-tiny.yml tests/modules/programs/vivid/themes/tiny.yml`

Still the same error 🤔 
```log
ℹ️ Discovering tests...
ℹ️ Running 1 test(s)...

--- Running test 1/1: test-vivid-example-config ---
❌ Test failed: test-vivid-example-config
this derivation will be built:
  /nix/store/v80z08bc92axxp0q6ndbfl186hvvmayp-nmt-test-vivid-example-config.drv
building '/nix/store/v80z08bc92axxp0q6ndbfl186hvvmayp-nmt-test-vivid-example-config.drv'...
nmt-test-vivid-example-config> Expected home-files/.config/vivid/themes/tiny.yml to be same as /nix/store/fnzcfj3vxnndbdnl15mcb117ilzyms2i-tiny.yml but were different:
nmt-test-vivid-example-config> --- actual
nmt-test-vivid-example-config> +++ expected
nmt-test-vivid-example-config> @@ -1 +1 @@
nmt-test-vivid-example-config> -{"colors":{"primary":"00aaff","secondary":"ff00aa"},"core":{"directory":{"foreground":"primary"},"executable-file":{"font-style":"bold","foreground":"secondary"}}}
nmt-test-vivid-example-config> \ No newline at end of file
nmt-test-vivid-example-config> +{"colors":{"primary":"00aaff","secondary":"ff00aa"},"core":{"directory":{"foreground":"primary"},"executable-file":{"font-style":"bold","foreground":"secondary"}}}
note: keeping build directory '/tmp/nix-build-nmt-test-vivid-example-config.drv-5/build'
error: builder for '/nix/store/v80z08bc92axxp0q6ndbfl186hvvmayp-nmt-test-vivid-example-config.drv' failed with exit code 1;
       last 7 log lines:
       > Expected home-files/.config/vivid/themes/tiny.yml to be same as /nix/store/fnzcfj3vxnndbdnl15mcb117ilzyms2i-tiny.yml but were different:
       > --- actual
       > +++ expected
       > @@ -1 +1 @@
       > -{"colors":{"primary":"00aaff","secondary":"ff00aa"},"core":{"directory":{"foreground":"primary"},"executable-file":{"font-style":"bold","foreground":"secondary"}}}
       > \ No newline at end of file
       > +{"colors":{"primary":"00aaff","secondary":"ff00aa"},"core":{"directory":{"foreground":"primary"},"executable-file":{"font-style":"bold","foreground":"secondary"}}}
       For full logs, run:
         nix log /nix/store/v80z08bc92axxp0q6ndbfl186hvvmayp-nmt-test-vivid-example-config.drv

ℹ️ Generated test directory at: /nix/store/4gh3zclg5w21h6hgggddn440ahnximwp-home-manager-generation/

--- Summary ---
❌ 1 of 1 test(s) failed:
  - test-vivid-example-config
```

aguirre-matteo commented:
Maybe `truncate --size=-1 ./tests/modules/programs/vivid/themes/tiny.yml` could work?

aguirre-matteo commented:
You didn't added a comment about why JSON instead of YAML but I guess there's no problem since anyone could just do a git blame and get to this PR.

khaneliman commented on modules/programs/vivid.nix:
```diff
@@ -34,8 +37,11 @@ in
+        "8-bit"
+        "24-bit"
+      ];
+      default = "24-bit";
```
Not a fan of changes that add maintenance burden for no real benefit. Also forces a cli flag instead of just going with the true app default.

ie: If they ever added different bit supports or changed their default, we'd be enforcing an enum that wouldn't allow a user to change and / or we'd enforce a default that conflicts with upstream.

khaneliman replied:
If you feel strong enough about wanting the enum back (assuming for documentation / guiding a user), I'd prefer to keep the `nullOr` for backwards compatibility and to decide whether to pass the cli flag at least and allow using upstream default. Something we've done elsewhere for maximum flexibility is a `nullOr (either str enum[])` to guide towards known good values but allow someone to pass a value that could be acceptable in the future.

arunoruto replied:
I will implement it tomorrow then!
And thanks for the hint about `nullOr (either str enum[])`, wasn't aware of it, but I will keep it in mind for the future!
So a potential implementation would look like this:
```nix
type =
  with types;
  nullOr (either str enum [
    "8-bit"
    "24-bit"
  ]);
```

trueNAHO replied:
> Something we've done elsewhere for maximum flexibility is a `nullOr (either str enum[])` to guide towards known good values but allow someone to pass a value that could be acceptable in the future.

Thanks for sharing! This is a great pattern.

trueNAHO reviewed (approved):
Since it has been a while since I reviewed this PR, I only vaguely remember the discussion. Is the following summary correct:

> This PR should:
>
> 1. Fix the YAML conversion by replacing it with JSON.
>
> -- https://github.com/nix-community/home-manager/pull/7996#discussion_r2435690017

This is covered by patch `[PATCH 2/3] vivid: replace theme content from yaml with json`.

> 2. Implement a second enabling mode for setting `LS_COLORS` when having set `cfg.activeTheme` to a static YAML attribute set. This second enabling mode should be toggable with something like `cfg.enableSessionVariables`, just like `cfg.enable`.
>
> -- https://github.com/nix-community/home-manager/pull/7996#discussion_r2435690017

IIUC, semi-statically evaluating `vivid generate` and preventing `pkgs.vivid` from becoming a runtime dependency by declaring only `home.sessionVariables.LS_COLORS` when `cfg.enableSessionVariables` is enabled, is postponed to a follow-up PR.

> Now, it also further restricts the `cfg.colorMode` type.
>
> -- https://github.com/nix-community/home-manager/pull/7996#discussion_r2435690017

This is covered by patch `[PATCH 3/3] vivid: change colorMode type to enum`.

---

In that case, the only relevant patch for Stylix would be `[PATCH 2/3] vivid: replace theme content from yaml with json`. I would really like to see (2) become a reality, especially since (IIRC) a working prototype exists.

trueNAHO commented on modules/programs/vivid.nix:
```diff
@@ -34,8 +37,15 @@ in
+            "24-bit"
+          ])
+        );
+      default = "24-bit";
```
The default should stay `null` to use the CLI's true default:

```suggestion
      default = null;
```
